File: ElasticModellingApporach/GA/individual.m
classdef individual
    %individual class. An indivudual is a single member of a population

    properties (Access = private)
        Field double;
        Fitness double;
    end

    methods (Access = public)

        %% Constructor
        function obj = individual(world)

            % constructs an instance of individual, and injects the world
            % object
            obj.Field = world.getField();               % init the Field variable
            obj = randomise(obj);                       % randomise the filed (initial guess)
            obj = determineFitness(obj, world);         % measure the fitness of this solution

        end

        %% Accessors
        function Field = getField(obj)
            Field = obj.Field;
        end

        function Fitness = getFitness(obj)
            Fitness = obj.Fitness;
        end

        function obj = updateField(obj,Field,world)

            obj.Field = Field;                          % Update Field
            obj = determineFitness(obj, world);         % measure the fitness of this solution
        end


    end

    methods (Access = private)

        function obj = randomise(obj)

            % Define limits
            lower_limit = -0.0025;  
            upper_limit = 0.0025; 
            
            % Randomize a and b between the limits
            a = lower_limit + (upper_limit - lower_limit) * rand;
            b = lower_limit + (upper_limit - lower_limit) * rand;
            
            % Create the vector x
            obj.Field = [a, b];
        end


        function obj = determineFitness(obj, world)

            % inject new randomised field into world
            world = world.updateField(obj.Field);

            % obtain resultant positions
            [DesiredXPos,DesiredYPos,xPos,yPos] = world.getPositions();

            % determine fintess
            obj.Fitness = ((xPos - DesiredXPos).^2 + (yPos - DesiredYPos).^2);

        end

    end

end

File: ElasticModellingApporach/cpp implementation/HomogenousFields/HomogenousFields/Individual.cpp
// Individual class. An individual is a single member of a population


#include "Individual.h"


// Constructor
Individual::Individual(World world) : world_(world) {

	Individual::Randomise();
	Individual::DetermineFitness();
}


// Destructor
Individual::~Individual() {

}

// Update the field and recalculate tentacle/ soultion fitness
void Individual::UpdateField(double Bx, double  Bz) {

	Bx_ = Bx;
	Bz_ = Bz;
	Individual::DetermineFitness();
}

// Accessor to return (a reference to) the fitness
double& Individual::GetFitness(){

	return fitness_;
}

void Individual::Randomise() {

	// Seed random number generator
	std::srand(std::time(0));

	// define the limits
	double lowerLimit = -0.0018;
	double upperLimit = 0.0018;

	// Randomize a and b between the limits
	Bx_ = lowerLimit + (upperLimit - lowerLimit) * (std::rand() / (double)RAND_MAX);
	Bz_ = lowerLimit + (upperLimit - lowerLimit) * (std::rand() / (double)RAND_MAX);
}

// Method to determine the fitness fo this solution
void Individual::DetermineFitness() {

	// Update the field
	world_.UpdateField(Bx_, Bz_);

	// get the desired and actual positions
	const auto& CurrentX = world_.GetCurrentX();
	const auto& CurrentZ = world_.GetCurrentZ();
	const auto& DesX = world_.GetDesiredX();
	const auto& DesZ = world_.GetDesiredZ();

	// Calculate difference between current and desired positions
	Eigen::VectorXd diffX = CurrentX - DesX;
	Eigen::VectorXd diffZ = CurrentZ - DesZ;

	// Concatenate X and Z differences
	Eigen::VectorXd diff;
	diff.resize(diffX.size() + diffZ.size());
	diff << diffX, diffZ;

	// Compute the norm of the concatenated difference vector and square it --- vector norm
	fitness_ =  std::pow(diff.norm(), 2);
}

// Accessor to return the fields
std::vector<double> Individual::GetField() {

	std::vector<double> Fields;

	Fields[0] = Bx_;
	Fields[1] = Bz_;

	return Fields;
}
